### PR TITLE
Create CustomIcon as suggested in #228

### PR DESCRIPTION
**Pull request by BibMartin**
Using a refactoring with ImageOverlay, you can provide an image url, file, or array.

### CODE REVIEW

File: folium/features.py [hunk 1]
```diff
@@ -7,10 +7,9 @@
 """
 from jinja2 import Template
 import json
-import base64
 
 from .utilities import color_brewer, _parse_size, legend_scaler, _locations_mirror, _locations_tolist, write_png,\
-    mercator_transform
```
Comment (ocefpaf) [inline folium/features.py]: We needed this re-factoring!  Thanks!!

@andrewgiessel this is your baby.  Can take a quick look at it before we merge this?

File: folium/features.py [hunk 2]
```diff
@@ -610,3 +585,60 @@ def __init__(self, image, bounds, opacity=1., attribution=None, origin='upper',
                     ).addTo({{this._parent.get_name()}});
             {% endmacro %}
             """)
+
+class CustomIcon(Icon):
```
Comment (ocefpaf) [inline folium/features.py]: I feel so ashamed for wasting an afternoon doing [this](https://github.com/ocefpaf/secoora_assets_map/blob/gh-pages/assets_map.ipynb) :flushed: 

This class is pure art!

**Comment by ocefpaf** [comment]
Beautiful!

I just re-wrote my horrible hack using this PR.

http://nbviewer.ipython.org/github/ocefpaf/secoora_assets_map/blob/gh-pages/test_geojson.ipynb

**Comment by andrewgiessel** [comment]
Yes, I'm on it! I have that and the retina PR at least

File: folium/utilities.py [hunk 3]
```diff
@@ -363,6 +365,51 @@ def mercator_transform(data, lat_bounds, origin='upper', height_out=None):
...
+            fileformat = image.name.lower().split('.')[-1]
+        else:
+            fileformat = 'png'
+        url = "data:image/{};base64,{}".format(fileformat,
+                                               base64.b64encode(image.read()).decode('utf-8'))
+    elif (not (isinstance(image,text_type) or isinstance(image,binary_type))) and hasattr(image,'__iter__'):
+        # We got an array-like object
+        if mercator_project:
+            data = mercator_transform(image,
+                                      [bounds[0][0], bounds[1][0]],
```
Comment (andrewgiessel) [inline folium/utilities.py]: bounds is not defined here- needs to be passed into from the __init__ method of ImageOverlay

**Comment by andrewgiessel** [comment]
This is amazing work!  Other than the bounds issue I did a line comment on above, it seems like a very nice refactoring.

Do the image_overlay tests still pass?  Do we more stuff there?

**Comment by BibMartin** [comment]
> bounds is not defined here- needs to be passed into from the init method of ImageOverlay

Silly me ; corrected.

> Do the image_overlay tests still pass? Do we more stuff there?

The test still passes ; I did not use `mercator_tranform` into it.
I'll add another test asap.

**Comment by ocefpaf** [comment]
@andrewgiessel Are you OK with @BibMartin's two extra commits?

**Comment by andrewgiessel** [comment]
Looks great to me, let's merge

**Comment by ocefpaf** [comment]
> Looks great to me, let's merge

Done!

**Comment by Dhanesh-raj** [comment]
import os
from io import BytesIO

style_function = lambda feature: dict(fillColor='#DEFFDE',
                                      color='#AECCAE',
                                      weight=1,
                                      opacity=1)
data_cm = kar_latlong
geojson = folium.GeoJson(data=country_shapes,
                         smooth_factor=2,
               style_function=lambda x: {'color':'black','fillColor':'green','weight':1},
                tooltip=folium.GeoJsonTooltip(fields=['name'],
                                              labels=False,
                                              sticky=True),
              highlight_function=lambda x: {'weight':3,'fillColor':'grey'},
                       ).add_to(world_map)


for i in range(0,len(data_cm)):
    icon_image = BytesIO(base64.b64decode(base64.b64encode(open(os.path.join(os.getcwd(),"./icons/",data_cm.iloc[i]['CountryName']+".png"), 'rb').read())))
    marker_icon = copy.copy(icon_image)
    icon = CustomIcon(
        marker_icon,
        icon_size=(30, 30),
        icon_anchor=(15, 15),
        popup_anchor=(-3, -76)
    )
    folium.Marker(
    location=[data_cm.iloc[i]['Latitude'], data_cm.iloc[i]['Longitude']],
    icon = icon,
    radius=float(15000.0),
    popup=folium.Popup(str(data_cm.iloc[i]['CountryName']))).add_to(world_map)

folium.LayerControl(collapsed=True).add_to(world_map)
world_map.save('icon.html')
world_map



**---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-42-16a4128b204a> in <module>
     20     icon_image = BytesIO(base64.b64decode(base64.b64encode(open(os.path.join(os.getcwd(),"./icons/",data_cm.iloc[i]['CountryName']+".png"), 'rb').read())))
     21     marker_icon = copy.copy(icon_image)
---> 22     icon = CustomIcon(
     23         marker_icon,
     24         icon_size=(30, 30),

NameError: name 'CustomIcon' is not defined**

**Pull request merged**